### ISSUE TITLE
Assume during assignments for non-nullable reference types

### ISSUE THREAD

**Issue by qubitz** [fernandoescolar/Xunit.Assume#1]
`Assume.That(...)` works well for checking conditions after uncertain operations during test setup, but it doesn't work with expressions that assign non-nullable reference types. For example:
```
string value = CouldBeNull() // Warning: Possible null reference assignment
```
something like
```
string value = CouldBeNull() ?? throw new AssumeException("Can not be null");
```
would be ideal, however `AssumeException` is an internal member.

Should I submit a PR changing `AssumeException` to be `public`? If not, what are some other alternatives?

P.S. Thanks for the library 👍

**Comment by fernandoescolar** [comment]
What do you think about adding a feature like this?

```csharp
string value = CouldBeNull() ?? Assume.Reject("Can not be null");
```

**Comment by qubitz** [comment]
That'll do 😊

**Comment by fernandoescolar** [comment]
I have added it in v 1.0.1

**Comment by qubitz** [comment]
Thanks for the fast response @fernandoescolar. I was under the assumption that Assume.Reject() would return an exception to allow for:
```c#
string value = CouldBeNull() ?? Assume.Reject("Can not be null");
```
However, with v1.0.1 Reject() returns void disallowing the above usage. I'd like to reopen the issue unless the return type was a typo rather than a design choice.

**Comment by fernandoescolar** [comment]
My mistake. 

I'm thinking that this problem could be solved in different ways:

- getting the return type as a generic type:
```csharp
string value = CouldBeNull() ?? Assume.Reject<string>("Can not be null");
```

- getting the return type as a parameter:
```csharp
string value = CouldBeNull() ?? Assume.Reject(CouldBeNull(), "Can not be null");
```

- or even returning the parameter value if it is not null:
```csharp
string value = Assume.NotNull(CouldBeNull(), "Can not be null");
```

**Comment by qubitz** [comment]
`Assume.Reject<string>()` seems the best option.

As a side note, what is the hesitancy against publicizing the AssumeException constructor? Throwing an exception after the `??` operator is very readable given how common its usage.

**Comment by fernandoescolar** [comment]
I think that throwing an AssumeException and interpret it as a test Skip is an internal behavior of the library. And I think it should not be exposed outside.

Your issue has opened my mind and I've been thinking of different ways to solve it. So I've just published a new version (v2.0.0) full of new stuff like this:

```csharp
var value = CouldBeNull() ?? Assume.Reject<string>("Can not be null");
```

```csharp
var value = Assume.NotNull(CouldBeNull(), "Can not be null");
```

```csharp
var value = CouldBeNull().AssumeNotNull("Can not be null");
```

```csharp
var value = CouldBeNull().Assume()
                         .NotNull("Can not be null");
```

**Comment by qubitz** [comment]
Just gave the new assumption APIs a try, works great 👍. Thanks for the quick turnaround!
